### PR TITLE
Add component-based demand forecasting notebooks

### PR DESCRIPTION
**Pull request by vlbejan**
# Description
Adding notebooks to showcase component-based approach (v2) to demand forecasting for single-, many models and HTS scenarios. 

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by ctian-msft — APPROVED** [review]

**Review by RSavage2 — APPROVED** [review]

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb [hunk 1]
```diff
@@ -0,0 +1,1339 @@
...
+    "print(f\"Data contains {nseries} individual time-series\\n---\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df[hierarchy_column_names].drop_duplicates()"
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:255]: Do we need this cell?

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb [hunk 2]
```diff
@@ -0,0 +1,1339 @@
...
+    "    allowed_training_algorithms=[\"LightGBM\"],\n",
+    "    iterations=5,\n",
+    "    n_cross_validations=2,\n",
+    "    forecast_step=max_horizon,\n",
+    "    track_child_runs=False,\n",
+    "    enable_early_stopping=True,\n",
+    "    hierarchy_training_level=hierarchy_training_level,\n",
+    "    hierarchy_column_names=hierarchy_column_names,\n",
+    ")\n",
+    "print(json.dumps(automl_settings, indent=4), \"\\n---\")"
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:483]: Do we need this line?

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb [hunk 3]
```diff
@@ -0,0 +1,1339 @@
...
+    "baseline_settings.update(\n",
+    "    {\"allowed_training_algorithms\": [\"Naive\"], \"n_cross_validations\": 2}\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Similarly to what we have done in section 6.3, we save the baseline experiment settings to as a JSON file."
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:747]: YAML file

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 4]
```diff
@@ -0,0 +1,1477 @@
...
+   "outputs": [],
+   "source": [
+    "metrics_automl = (\n",
+    "    pd.Series(metrics_automl_series).to_frame(name=\"score\").reset_index(drop=False)\n",
+    ")\n",
+    "metrics_base = (\n",
+    "    pd.Series(metrics_base_series).to_frame(name=\"score\").reset_index(drop=False)\n",
+    ")\n",
+    "metrics_all = pd.DataFrame(\n",
+    "    [metrics_automl_series, metrics_base_series], index=[\"score_automl\", \"score_base\"]\n",
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:1068]: Let us shorten:
metrics_all = pd.DataFrame([metrics_automl_series, metrics_base_series], index=['score_automl', 'score_base']).T

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 5]
```diff
@@ -0,0 +1,1477 @@
...
+    "baseline_settings.update(\n",
+    "    {\"allowed_training_algorithms\": [\"Naive\"], \"n_cross_validations\": 2}\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Similarly to what we have done in section 6.3, we save the baseline experiment settings to as a JSON file."
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:890]: to a YAML file

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 6]
```diff
@@ -0,0 +1,1477 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_folder_and_save_as_parquet(file_uri, output_folder):\n",
+    "    os.makedirs(output_folder, exist_ok=True)\n",
+    "    data_frame = pd.read_csv(file_uri, parse_dates=[time_column_name])\n",
+    "    file_name = file_uri.split(\"/\")[-1].replace(\".csv\", \".parquet\")\n",
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:246]: Let us use library function here:

```python
file_name = os.path.splitext(os.path.split(file_uri)[-1])[0] + ".parquet"
```

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 7]
```diff
@@ -0,0 +1,1477 @@
...
+    "    data_path = os.path.join(output_folder, file_name)\n",
+    "    data_frame.to_parquet(data_path, index=False)\n",
+    "    return None\n",
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:249]: As far as you still will read data frame in the next cell, let us return the data frame.
```python
return data_frame
```

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
+import json
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:4]: We do not use json here.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import os
+import re
+import shutil
+
+import pandas as pd
+
+from typing import List, Union, Tuple
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:11]: Second import of typing

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
+import json
+import os
+import re
+import shutil
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:7]: Unused import

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
+import json
+import os
+import re
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:6]: Unused import

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
+import json
+import os
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:5]: unused import

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:3]: unused import

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/scripts/helper_scripts.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Dict, Optional, List
+
+import argparse
```
Comment (nick863) [inline sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/scripts/helper_scripts.py:3]: Please see the comments for the many model notebook helper script.

**Review by nick863 — CHANGES_REQUESTED** [review]

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 15]
```diff
@@ -0,0 +1,1477 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Automated Machine Learning\n",
+    "**Demand Forecasting Using Many Models**\n",
+    "\n",
+    "## Contents\n",
+    "1. [Introduction](#Introduction)\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:20]: You are correct. None of these links work, including ALL of our notebooks both v1 and v2. Will issue a separate PR fixing this issue in other notebooks.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb [hunk 16]
```diff
@@ -0,0 +1,1477 @@
...
+    "1. [Download Output](#DownloadOutput)\n",
+    "1. [Compare Evaluation Results](#CompareResults)\n",
+    "1. [Deployment](#Deployment)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Introduction\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:36]: done

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb [hunk 17]
```diff
@@ -0,0 +1,1339 @@
...
+    "from azure.ai.ml.constants import AssetTypes\n",
+    "from azure.ai.ml.dsl import pipeline\n",
+    "from azure.ai.ml.entities import (\n",
+    "    BatchEndpoint,\n",
+    "    BatchDeployment,\n",
+    "    AmlCompute,\n",
+    "    PipelineComponentBatchDeployment,\n",
+    ")\n",
+    "\n",
+    "from azure.ai.ml.entities._deployment.job_definition import JobDefinition\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:88]: done.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb [hunk 18]
```diff
@@ -0,0 +1,1339 @@
...
+    "1. [Download Output](#DownloadOutput)\n",
+    "1. [Compare Evaluation Results](#CompareResults)\n",
+    "1. [Deployment](#Deployment)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Introduction\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:36]: done

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb (hunk 1 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:255]: This is by design.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb (hunk 2 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/automl-forecasting-demand-hierarchical-timeseries-in-pipeline.ipynb:483]: This is by design.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py (hunk 8 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:4]: removed

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py (hunk 9 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:11]: fixed

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py (hunk 10 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/scripts/helper_scripts.py:7]: removed

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/scripts/helper_scripts.py (hunk 14 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-hierarchical-timeseries-in-pipeline/scripts/helper_scripts.py:3]: Fixed.

File: sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb (hunk 6 above)
Comment (vlbejan) [reply sdk/python/jobs/pipelines/1k_demand_forecasting_with_pipeline_components/automl-forecasting-demand-many-models-in-pipeline/automl-forecasting-demand-many-models-in-pipeline.ipynb:246]: fixed

**Review by nick863 — APPROVED** [review]

**Pull request merged**